OK, maybe the `if abs(x)<1e15` criteria should be used to allow `as_int(1e3) -> 1000`. If the number is small enough then all known information about the number allows it to be distinguished between an integer and a non-integer. If n is a Float, then to the given precision if `int(n) == n` then we can conclude the same: to the degree of precision given, the number is indistinguishable from an integer. e.g.

```python
>>> n = Float('1.00000000001', 2)
>>> int(n) == n
True
```

Here are the errors from the Travis run:
```
___________ sympy/core/tests/test_evalf.py:test_evalf_integer_parts ____________
Traceback (most recent call last):
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/sympy-1.4.dev0-py2.7.egg/sympy/utilities/runtests.py", line 1306, in _timeout
    function()
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/sympy-1.4.dev0-py2.7.egg/sympy/core/tests/test_evalf.py", line 258, in test_evalf_integer_parts
    assert ceiling(x).evalf(subs={x: 2. + 3*I}) == 2 + 3*I
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/sympy-1.4.dev0-py2.7.egg/sympy/core/evalf.py", line 1437, in evalf
    result = evalf(self, prec + 4, options)
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/sympy-1.4.dev0-py2.7.egg/sympy/core/evalf.py", line 1307, in evalf
    r = rf(x, prec, options)
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/sympy-1.4.dev0-py2.7.egg/sympy/core/evalf.py", line 408, in evalf_ceiling
    return get_integer_part(expr.args[0], 1, options)
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/sympy-1.4.dev0-py2.7.egg/sympy/core/evalf.py", line 398, in get_integer_part
    re_, re_acc = calc_part(re(expr, evaluate=False), ire)
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/sympy-1.4.dev0-py2.7.egg/sympy/core/evalf.py", line 389, in calc_part
    raise PrecisionExhausted
PrecisionExhausted
________________________________________________________________________________
____________ sympy/discrete/tests/test_transforms.py:test_ntt_intt _____________
Traceback (most recent call last):
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/sympy-1.4.dev0-py2.7.egg/sympy/utilities/runtests.py", line 1306, in _timeout
    function()
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/sympy-1.4.dev0-py2.7.egg/sympy/discrete/tests/test_transforms.py", line 60, in test_ntt_intt
    assert ntt([1.0, 2.0, 3.0], p) == ntt([1, 2, 3], p)
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/sympy-1.4.dev0-py2.7.egg/sympy/discrete/transforms.py", line 232, in ntt
    return _number_theoretic_transform(seq, prime=prime)
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/sympy-1.4.dev0-py2.7.egg/sympy/discrete/transforms.py", line 146, in _number_theoretic_transform
    a = [as_int(x) % p for x in seq]
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/sympy-1.4.dev0-py2.7.egg/sympy/core/compatibility.py", line 416, in as_int
    raise ValueError('%s is not an integer' % (n,))
ValueError: 1.0 is not an integer
________________________________________________________________________________
___________ sympy/matrices/tests/test_matrices.py:test_creation_args ___________
Traceback (most recent call last):
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/sympy-1.4.dev0-py2.7.egg/sympy/utilities/runtests.py", line 1306, in _timeout
    function()
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/sympy-1.4.dev0-py2.7.egg/sympy/matrices/tests/test_matrices.py", line 1619, in test_creation_args
    assert zeros(3.) == zeros(3)
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/sympy-1.4.dev0-py2.7.egg/sympy/matrices/dense.py", line 1491, in zeros
    return Matrix.zeros(*args, **kwargs)
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/sympy-1.4.dev0-py2.7.egg/sympy/matrices/common.py", line 903, in zeros
    rows, cols = as_int(rows), as_int(cols)
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/sympy-1.4.dev0-py2.7.egg/sympy/core/compatibility.py", line 416, in as_int
    raise ValueError('%s is not an integer' % (n,))
ValueError: 3.0 is not an integer
________________________________________________________________________________
_____________ sympy/core/tests/test_numbers.py:test_invert_numbers _____________
Traceback (most recent call last):
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/sympy-1.4.dev0-py2.7.egg/sympy/utilities/runtests.py", line 1306, in _timeout
    function()
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/sympy-1.4.dev0-py2.7.egg/sympy/core/tests/test_numbers.py", line 1741, in test_invert_numbers
    assert S(2).invert(5.) == 3
AssertionError
________________________________________________________________________________
__________________ sympy/core/tests/test_subs.py:test_powers ___________________
Traceback (most recent call last):
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/sympy-1.4.dev0-py2.7.egg/sympy/utilities/runtests.py", line 1306, in _timeout
    function()
  File "/home/travis/virtualenv/python2.7.14/lib/python2.7/site-packages/sympy-1.4.dev0-py2.7.egg/sympy/core/tests/test_subs.py", line 81, in test_powers
    assert (x**(4.0*y)).subs(x**(2.0*y), n) == n**2.0
AssertionError
```

--- a/sympy/core/evalf.py
+++ b/sympy/core/evalf.py
@@ -367,12 +367,14 @@ def get_integer_part(expr, no, options, return_ints=False):
             if s:
                 doit = True
                 from sympy.core.compatibility import as_int
+                # use strict=False with as_int because we take
+                # 2.0 == 2
                 for v in s.values():
                     try:
-                        as_int(v)
+                        as_int(v, strict=False)
                     except ValueError:
                         try:
-                            [as_int(i) for i in v.as_real_imag()]
+                            [as_int(i, strict=False) for i in v.as_real_imag()]
                             continue
                         except (ValueError, AttributeError):
                             doit = False

--- a/sympy/core/power.py
+++ b/sympy/core/power.py
@@ -671,7 +671,7 @@ class Pow(Expr):
                     # Allow fractional powers for commutative objects
                     pow = coeff1/coeff2
                     try:
-                        pow = as_int(pow)
+                        pow = as_int(pow, strict=False)
                         combines = True
                     except ValueError:
                         combines = isinstance(Pow._eval_power(

--- a/sympy/core/tests/test_numbers.py
+++ b/sympy/core/tests/test_numbers.py
@@ -1738,9 +1738,9 @@ def test_issue_10020():
 def test_invert_numbers():
     assert S(2).invert(5) == 3
     assert S(2).invert(S(5)/2) == S.Half
-    assert S(2).invert(5.) == 3
+    assert S(2).invert(5.) == 0.5
     assert S(2).invert(S(5)) == 3
-    assert S(2.).invert(5) == 3
+    assert S(2.).invert(5) == 0.5
     assert S(sqrt(2)).invert(5) == 1/sqrt(2)
     assert S(sqrt(2)).invert(sqrt(3)) == 1/sqrt(2)
 

--- a/sympy/discrete/tests/test_transforms.py
+++ b/sympy/discrete/tests/test_transforms.py
@@ -56,8 +56,7 @@ def test_ntt_intt():
     raises(ValueError, lambda: ntt([1.2, 2.1, 3.5], p))
     raises(ValueError, lambda: ntt([3, 5, 6], q))
     raises(ValueError, lambda: ntt([4, 5, 7], r))
-
-    assert ntt([1.0, 2.0, 3.0], p) == ntt([1, 2, 3], p)
+    raises(ValueError, lambda: ntt([1.0, 2.0, 3.0], p))
 
 
 def test_fwht_ifwht():

--- a/sympy/matrices/tests/test_matrices.py
+++ b/sympy/matrices/tests/test_matrices.py
@@ -1612,7 +1612,7 @@ def test_creation_args():
     raises(TypeError, lambda: zeros(1, 2, 3, 4))
     assert zeros(long(3)) == zeros(3)
     assert zeros(Integer(3)) == zeros(3)
-    assert zeros(3.) == zeros(3)
+    raises(ValueError, lambda: zeros(3.))
     assert eye(long(3)) == eye(3)
     assert eye(Integer(3)) == eye(3)
     assert eye(3.) == eye(3)

--- a/sympy/matrices/tests/test_matrices.py
+++ b/sympy/matrices/tests/test_matrices.py
@@ -1615,7 +1615,7 @@ def test_creation_args():
     raises(ValueError, lambda: zeros(3.))
     assert eye(long(3)) == eye(3)
     assert eye(Integer(3)) == eye(3)
-    assert eye(3.) == eye(3)
+    raises(ValueError, lambda: eye(3.))
     assert ones(long(3), Integer(4)) == ones(3, 4)
     raises(TypeError, lambda: Matrix(5))
     raises(TypeError, lambda: Matrix(1, 2))

--- a/sympy/ntheory/primetest.py
+++ b/sympy/ntheory/primetest.py
@@ -495,10 +495,30 @@ def isprime(n):
     Notes
     =====
 
-    This routine is not meant to deal with expressions which may
-    represent a prime number. Floats -- even those that are
-    equal to an integer value -- return False because they
-    represent numbers of limited precision.
+    This routine is intended only for integer input, not numerical
+    expressions which may represent numbers. Floats are also
+    rejected as input because they represent numbers of limited
+    precision. While it is tempting to permit 7.0 to represent an
+    integer there are errors that may "pass silently" if this is
+    allowed:
+
+    >>> from sympy import Float, S
+    >>> int(1e3) == 1e3 == 10**3
+    True
+    >>> int(1e23) == 1e23
+    True
+    >>> int(1e23) == 10**23
+    False
+
+    >>> near_int = 1 + S(1)/10**19
+    >>> near_int == int(near_int)
+    False
+    >>> n = Float(near_int, 10)  # truncated by precision
+    >>> n == int(n)
+    True
+    >>> n = Float(near_int, 20)
+    >>> n = int(n)
+    False
 
     See Also
     ========

